### PR TITLE
Improve functionality for Peloton App users

### PR DESCRIPTION
**Pull request by grantland**
* Cadence and Resistance Tiles - Improve readability for bikes using red, limegreen, and orange for < min, <= max, < max respectively
* Peloton Offset Tile
  * Rename to Peloton Remaining/Rem.
  * Add remaining time to make it easier to match with Peloton app
  * Swap remaining time with offset as remaining time is more useful
* Time to Next Tile
  * Add Current Row Elapsed Time as secondary text
  * Enable writing to skip forward/back to next/previous row

Note: Time To Next -/+ along with Peloton Remaining makes it super easy to catch up to the correct timestamp now!

Screenshot_1645774076

### CODE REVIEW

**Comment by cagnulein** [comment]
hi @grantland ! Thanks for the contribution. I see that there is some syntax error on it. Did you test this?

**Comment by grantland** [comment]
Sorry, I haven't been able to test these changes since I've been unable to get my QT dev environment working, although I think I'm getting close!

**Comment by cagnulein** [comment]
@grantland are you building for linux, android or iOS? If you need help please let me know. I'm waiting your test before merging this, just to be sure ;)

**Comment by grantland** [comment]
I'm trying to build for Android using a Mac dev env. Right now I think my main issue is that I was trying to use the latest QT version (6.2.3) but it seems like this project requires 5.15.2. Correct me if this is wrong but I'm downloading/installing that now and will probably try to continue tomorrow as it'll take ~2 hours.

Also I've never used QT before, but my prior issues if anyone else runs into this:
* Unknown module(s) in QT...  -> Install additional libraries
* "" architectures defined in kit -> Not sure what happened here but messing with kits, device, etc. might have somehow fixed this

Also while I have you, is there any reason Android releases seem to be behind iOS? iOS App Store version seems to be at 2.9 while Android Google Play version seems to be at 2.8.142. Only reason is that I noticed that my wife's version (iOS) was slightly different than mine (Android).

And thanks so much for building this awesome software and making it open source! I'll also be looking to implement the following to improve our Echelon EX5 + Peloton app usage:
* peloton resistance - add setting to follow % between min/max (i.e. min/max of 55/65 with setting of 60 is 50% so next step with min/max of 30/35 auto updates resistance to 32)
* peloton offset - show time remaining to easily match peloton app
* current step button - writable to allow jumping to next step to easily match instructor cues

**Comment by cagnulein** [comment]
> I'm trying to build for Android using a Mac dev env. Right now I think my main issue is that I was trying to use the latest QT version (6.2.3) but it seems like this project requires 5.15.2. Correct me if this is wrong but I'm downloading/installing that now and will probably try to continue tomorrow as it'll take ~2 hours.

Yes currently it's build with a custom version of the 5.15.2 but you can use the standard 5.15.2
If you want you can open a PR for a Qt6. It's something that i would like to try myself too

> 
> Also I've never used QT before, but my prior issues if anyone else runs into this:
> 
> * Unknown module(s) in QT...  -> Install additional libraries
> * "" architectures defined in kit -> Not sure what happened here but messing with kits, device, etc. might have somehow fixed this
> 
> Also while I have you, is there any reason Android releases seem to be behind iOS? iOS App Store version seems to be at 2.9 while Android Google Play version seems to be at 2.8.142. Only reason is that I noticed that my wife's version (iOS) was slightly different than mine (Android).

That's because i'm updating a version on the store ONLY if really needs to :) you can join the beta from the play store page to gain access to the latest one ;)

> 
> And thanks so much for building this awesome software and making it open source! I'll also be looking to implement the following to improve our Echelon EX5 + Peloton app usage:
> 
> * peloton resistance - add setting to follow % between min/max (i.e. min/max of 55/65 with setting of 60 is 50% so next step with min/max of 30/35 auto updates resistance to 32)
> * peloton offset - show time remaining to easily match peloton app
> * current step button - writable to allow jumping to next step to easily match instructor cues

Wonderful!

**Comment by grantland** [comment]
I'm now getting errors when trying to run clean master without my changes and I'm not sure if it's something wrong with my build env: https://gist.github.com/grantland/56aab09c37dea2ae3d601c911409e876

I've tried both the default Android Qt 5.12.2 Clang arm64-v8a and armeabi-v7a

> Yes currently it's build with a custom version of the 5.15.2 but you can use the standard 5.15.2
If you want you can open a PR for a Qt6. It's something that i would like to try myself too

Could this be an issue? Would it be possible for me to use this custom version as well?

It'd be great to know if there's anything I missed when setting up my env and lmk if there's any more info that could help debug this!

**Comment by cagnulein** [comment]
@grantland i saw a notification from you about this by email but i can't see it on Github. Did you delete it? Do you need help? Let me know

**Comment by grantland** [comment]
Yes, I deleted it because the issue I was having was due to my user error somehow misreading 5.12.2 for 5.15.2 which was causing me runtime errors.

I ended up getting everything running, but now I might not be able to test until next week due to being away from bike unless there are some ways to test without hardware.

**Comment by cagnulein** [comment]
@grantland great! Yes you can test it enabling the "apple watch fake device in the experimental settings"! Let me know

**Comment by grantland** [comment]
* New code compiles but haven't tested color changing cadence and resistance as fake bike doesn't allow for cadence and resistance
* Tested all other functionality
* Included many feature changes/additions in this PR but each is split up into different commits. LMK if this is OK but I can also split this up into separate PRs as well!

**Comment by cagnulein** [comment]
thanks @grantland i didn't check all the code yet but it seems a good work! Could i ask you to recommit this in order to don't commit the unwanted files (all the iOS unchanged file shouldn't be in the commit). I don't know if i can exclude these from the PR

**Comment by grantland** [comment]
Cleaned up! I'll move that discussion to another PR :)

File: src/homeform.cpp [hunk 1]
```diff
@@ -1711,10 +1720,16 @@ void homeform::Minus(const QString &name) {
...
+            // Offset:
+            // 1. To account for current tick
+            // 2. To bounce back to previous
+            trainProgram->decreaseElapsedTime(QTime(0, 0, 0).secsTo(trainProgram->currentRowElapsedTime()) + 2);
+        }
     } else if (name.contains(QStringLiteral("peloton_offset"))) {
-
         if (bluetoothManager->device() && trainProgram) {
-            trainProgram->decreaseElapsedTime(1);
+            trainProgram->increaseElapsedTime(1);
```
Comment (cagnulein) [inline src/homeform.cpp:1732]: why did you change this? this will broke the compatibilty...

File: src/homeform.cpp [hunk 2]
```diff
@@ -1576,10 +1582,13 @@ void homeform::Plus(const QString &name) {
...
         }
+    } else if (name.contains(QStringLiteral("remainingtimetrainprogramrow"))) {
+        if (bluetoothManager->device() && trainProgram) {
+            trainProgram->increaseElapsedTime(QTime(0, 0, 0).secsTo(trainProgram->currentRowRemainingTime()));
+        }
     } else if (name.contains(QStringLiteral("peloton_offset"))) {
-
         if (bluetoothManager->device() && trainProgram) {
-            trainProgram->increaseElapsedTime(1);
+            trainProgram->decreaseElapsedTime(1);
```
Comment (cagnulein) [inline src/homeform.cpp:1591]: why did you change this? this will broke the compatibilty...

**Comment by grantland** [comment]
Removed +/- changes and changed "Peloton Offset" to "Peloton Remaining"/"Peloton Rem."

Screenshot_1645923372

File: src/homeform.cpp [hunk 3]
```diff
@@ -208,7 +208,7 @@ homeform::homeform(QQmlApplicationEngine *engine, bluetooth *bl) {
     steeringAngle = new DataObject(QStringLiteral("Steering"), QStringLiteral("icons/icons/cadence.png"),
                                    QStringLiteral("0"), false, QStringLiteral("steeringangle"), 48, labelFontSize);
     peloton_offset =
-        new DataObject(QStringLiteral("Peloton Offset"), QStringLiteral("icons/icons/clock.png"), QStringLiteral("0"),
+        new DataObject(QStringLiteral("Peloton Rem."), QStringLiteral("icons/icons/clock.png"), QStringLiteral("0"),
```
Comment (cagnulein) [inline src/homeform.cpp:211]: we can't change tile in this way, you have to keep in mind that thousand of users are you using this. Changing even something very little cause a lot of work for me to inform all the users. I would like to revert this.

File: src/settings.qml [hunk 4]
```diff
@@ -2404,7 +2404,7 @@ import Qt.labs.settings 1.0
 
                     AccordionCheckElement {
                         id: pelotonOffsetEnabledAccordion
-                        title: qsTr("Peloton Offset")
+                        title: qsTr("Peloton Remaining")
```
Comment (cagnulein) [inline src/settings.qml:2407]: same as above

File: src/homeform.cpp [hunk 5]
```diff
@@ -2210,21 +2224,44 @@ void homeform::update() {
...
                     this->target_peloton_resistance->setSecondLine(
                         QStringLiteral("MIN: ") +
-                        QString::number(trainProgram->currentRow().lower_requested_peloton_resistance, 'f', 0) +
+                        QString::number(lower_requested_peloton_resistance, 'f', 0) +
                         QStringLiteral(" MAX: ") +
-                        QString::number(trainProgram->currentRow().upper_requested_peloton_resistance, 'f', 0));
+                        QString::number(upper_requested_peloton_resistance, 'f', 0));
+
+                    if (peloton_resistance < lower_requested_peloton_resistance) {
+                        this->peloton_resistance->setValueFontColor(QStringLiteral("red"));
```
Comment (cagnulein) [inline src/homeform.cpp:2237]: i already made a poll about this on the facebook group. The result was mixed: some people want this and some people hate this :D so i guess we should put a setting for these tiles in order to enable the color thing (disabled by default). You can put the setting in the tiles setting (maybe we can put a "color" setting for each tile too). What do you think?

File: src/homeform.cpp [hunk 6]
```diff
@@ -1963,9 +1972,14 @@ void homeform::update() {
         pidHR->setValue(QString::number(treadmill_pid_heart_zone));
 
         if (trainProgram) {
-            peloton_offset->setValue(QString::number(trainProgram->offsetElapsedTime()) + QStringLiteral(" sec."));
+            peloton_offset->setValue(trainProgram->remainingTime().toString("h:mm:ss"));
+            peloton_offset->setSecondLine(QString::number(trainProgram->offsetElapsedTime()) + QStringLiteral(" sec."));
             remaningTimeTrainingProgramCurrentRow->setValue(
                 trainProgram->currentRowRemainingTime().toString(QStringLiteral("h:mm:ss")));
+            remaningTimeTrainingProgramCurrentRow->setSecondLine(
+                trainProgram->currentRowElapsedTime().toString(QStringLiteral("h:mm:ss")));
```
Comment (cagnulein) [inline src/homeform.cpp:1980]: i think for this we should make a video and show this to the facebook community. I would like to have a feedback from the users because it could change the current behaviour. What do you think?

**Comment by cagnulein** [comment]
hi @grantland did you read my previous comments?

**Comment by grantland** [comment]
@cagnulein yes I did, have you seen my responses?

**Comment by cagnulein** [comment]
@grantland Nope they are all 4 without answers:(

**Comment by grantland** [comment]
I only saw 2 and attempted to consolidate the conversation. LMK if there's anything else I'm missing!

**Comment by cagnulein** [comment]
@grantland I commented directly the code . I'm not on the pc so I can't link to them, but you can easily see them if you check the previous notifications of this thread. I will point you to them tomorrow. Thanks and sorry for the misunderstanding

**Review by grantland — COMMENTED** [review]
Turns out they were all pending and I never submitted! Sorry about that, I haven't used Github PRs in a bit.

File: src/homeform.cpp [hunk 7]
```diff
@@ -1963,9 +1972,14 @@ void homeform::update() {
         pidHR->setValue(QString::number(treadmill_pid_heart_zone));
 
         if (trainProgram) {
-            peloton_offset->setValue(QString::number(trainProgram->offsetElapsedTime()) + QStringLiteral(" sec."));
+            peloton_offset->setValue(trainProgram->remainingTime().toString("h:mm:ss"));
+            peloton_offset->setSecondLine(QString::number(trainProgram->offsetElapsedTime()) + QStringLiteral(" sec."));
```
[the remark this answers is not in the document]
Comment (cagnulein) [reply src/homeform.cpp:1976]: Yes I guess the best way to conserve the legacy is to create a new one. We need to play safely with changes because a lot of people are involved on this ;) thanks

File: src/homeform.cpp (hunk 5 above)
Comment (cagnulein) [reply src/homeform.cpp:2237]: Or we can just add a new "Color" setting for each tile so they can choose this behaviour for every tile. Default values could be the current behaviour. I just want to make sure that everyone will be happy about this

File: src/homeform.cpp (hunk 6 above)
Comment (cagnulein) [reply src/homeform.cpp:1980]: Yeah sorry, now I got your idea behind is!

**Comment by grantland** [comment]
* Reverted changes to Peloton Offset and instead created a new Peloton Remaining tile
* Added settings for enabling colors for Cadence and Peloton Resistance

**Comment by cagnulein** [comment]
thanks, since the #714 is partially also here, we should merge first that one ;)

**Comment by grantland** [comment]
sorry about that, still getting used to git again 😄 

split up the two PRs again

**Comment by cagnulein** [comment]
thanks, i'm gonna build the android and ios version with the profiles #714 and I soon I will have confirmation I will merge also this one in order to put not too much gas on the fire ;)
